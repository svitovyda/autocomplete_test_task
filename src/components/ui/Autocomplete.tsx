import debounce from 'debounce';
import * as React from 'react';

const Input: React.FC<React.InputHTMLAttributes<HTMLInputElement>> = (props) => {
  return <input {...props} />;
};

export interface AutocompleteProps<T> {
  onQueryChanged?: (query: string) => void;
  onItemSelected?: (item: T) => void;
  minAcceptableLength?: number;
  debounceInterval?: number;
  InputComponent?: React.ComponentType<React.InputHTMLAttributes<HTMLInputElement>>;
  placeholder?: string;
  maxOptionsToShow?: number;
  autoCalculate?: boolean;
  data: T[];
  dataLabel?: (item: T) => string;
  dataId?: (item: T) => string | number;
}
/*
 * This component renders input with autocompletion
 * It can be in two modes:
 * - autoCalculate ("static")
 *   it gets all the data once and then as user interacts, searches through this given data
 * - "dynamic"
 *   it assumes that all the calculations with data search by query happen outside of the component,
 *   and it just renders data
 */
export const Autocomplete = <T,>({
  onQueryChanged,
  onItemSelected,
  minAcceptableLength = 3,
  debounceInterval = 3,
  InputComponent = Input,
  placeholder = '',
  maxOptionsToShow = 5,
  autoCalculate,
  data,
  dataLabel = (item: T) => (item as any).toString(),
  dataId = (item: T) => (item as any).toString(),
}: AutocompleteProps<T>): JSX.Element => {
  const [input, setInput] = React.useState(''); // current state of the input
  const [query, setQuery] = React.useState(''); // deboused query
  const [labelsToShow, setLabelsToShow] = React.useState<string[]>([]);

  const onFocusIn = React.useCallback(() => {
    // show drop-out, data to show have to be set in `handleChange`
  }, []);

  const onFocusOut = React.useCallback(() => {
    // hide drop-out, do not clear query (?)
    setLabelsToShow([]);
  }, []);

  // eslint doesn't know how to work with debounce
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const updateSearchQuery = React.useCallback(
    debounce((value: string) => {
      if (value.length >= minAcceptableLength) {
        onQueryChanged?.(value);
        setQuery(value);
        // set data
      }
    }, debounceInterval),
    [onQueryChanged, minAcceptableLength, debounceInterval]
  );

  const handleChange = React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const { value } = e.target;
      setInput(value);
      updateSearchQuery(value);
    },
    [setInput, updateSearchQuery]
  );

  // to prevent compilation error, TODO: remove!
  console.log(onItemSelected, maxOptionsToShow);

  return (
    <InputComponent
      type="text"
      value={input}
      onChange={handleChange}
      placeholder={placeholder}
      onBlur={onFocusOut}
      onFocus={onFocusIn}
    />
  );
};
