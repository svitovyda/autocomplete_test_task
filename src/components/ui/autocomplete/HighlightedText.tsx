import type { QueryType } from '../../../utils/textSearch';
import React from 'react';
import styled from 'styled-components';

interface HighlightedTextProps {
  text: string;
  query: QueryType;
  caseSensitive?: boolean;
}

const Highlight = styled.span`
  background-color: yellow; /* or any other highlight color */
  font-weight: bold;
`;

export const HighlightedText: React.FC<HighlightedTextProps> = React.memo(
  ({ text, query, caseSensitive = false }: HighlightedTextProps) => {
    if (!query) return <>{text}</>;

    const regex = new RegExp(`(${query})`, caseSensitive ? 'g' : 'gi');
    const parts = text.split(regex);

    return (
      <>
        {parts.map((part, index) =>
          regex.test(part) ? <Highlight key={index}>{part}</Highlight> : part
        )}
      </>
    );
  }
);

HighlightedText.displayName = 'HighlightedText';
