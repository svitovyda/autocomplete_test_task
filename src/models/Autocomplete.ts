export type ItemID = string | number;

export interface ShowItem {
  id: ItemID;
  label: string;
}

export interface DictionaryItem<T> {
  itemToShow: ShowItem;
  itemData: T;
}
