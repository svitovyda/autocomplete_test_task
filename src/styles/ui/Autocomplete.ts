import styled from 'styled-components';

export const SelectContainer = styled.div`
  width: 100%;
  margin: 0;
  padding: 0;
  display: block;
  height: 0;
`;
