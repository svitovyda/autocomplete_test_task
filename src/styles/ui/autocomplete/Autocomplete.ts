import styled from 'styled-components';

export const AutocompleteInput = styled.input`
  display: inline-block;
  padding: 4px 2px;
  cursor: text;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #aaa;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin: 0;
  padding: 8px;
  outline-width: 0;
  color: inherit;
  ::placeholder {
    color: #aaa;
    font-size: 16px;
  }
`;
