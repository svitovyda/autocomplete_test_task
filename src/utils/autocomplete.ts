import type { DataArrayType } from '../hooks/autocomplete/useDataShow';
import type { DictionaryItem, ItemID } from '../models/Autocomplete';

export const eq = <T>(a: T): T => a;

export const defaultLabelId = <T>(item: T): string => {
  return (item as any).toString();
};

export const convertToShow = <T>(
  items: DataArrayType<T>,
  dataId: (item: T) => ItemID = defaultLabelId,
  dataLabel: (item: T) => string = defaultLabelId,
  max: number = 5
): DictionaryItem<T>[] =>
  items
    .slice(0, max)
    .map((item) => ({ itemData: item, itemToShow: { id: dataId(item), label: dataLabel(item) } }));
