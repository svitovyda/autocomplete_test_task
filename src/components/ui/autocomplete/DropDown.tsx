/* eslint-disable react/prop-types */
import type { DataArrayType } from '../../../hooks/autocomplete/useDataShow';
import type { DictionaryItem } from '../../../models/Autocomplete';
import {
  DropDownContainer,
  DropdownItem,
  DropdownList,
} from '../../../styles/ui/autocomplete/DropDown';
import type { QueryType } from '../../../utils/textSearch';
import { HighlightedText } from './HighlightedText';
import _ from 'lodash';
import React from 'react';

interface DropdownProps<T> {
  items: DataArrayType<DictionaryItem<T>>;
  query?: QueryType;
  caseSensitive?: boolean;
  onSelectItem: (item: DictionaryItem<T>) => void;
}

export const DropDown = React.memo(
  React.forwardRef<HTMLDivElement, DropdownProps<any>>(
    <T,>(
      { items, query, caseSensitive = false, onSelectItem }: DropdownProps<T>,
      ref: React.ForwardedRef<HTMLDivElement>
    ) => {
      return (
        <DropDownContainer ref={ref}>
          <DropdownList>
            {items.map((item) => (
              <DropdownItem key={item.itemToShow.id} onClick={() => onSelectItem(item)}>
                <HighlightedText
                  text={item.itemToShow.label}
                  query={query}
                  caseSensitive={caseSensitive}
                />
              </DropdownItem>
            ))}
          </DropdownList>
        </DropDownContainer>
      );
    }
  ),
  _.isEqual
);

DropDown.displayName = 'DropDown';
