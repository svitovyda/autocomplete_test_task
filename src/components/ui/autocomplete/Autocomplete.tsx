import { type DataArrayType, useDataShow } from '../../../hooks/autocomplete/useDataShow';
import { useInputQuery } from '../../../hooks/autocomplete/useInputQuery';
import { useClickOutside } from '../../../hooks/useClickOutside';
import type { DictionaryItem } from '../../../models/Autocomplete';
import { AutocompleteInput } from '../../../styles/ui/autocomplete/Autocomplete';
import { defaultLabelId } from '../../../utils/autocomplete';
import type { QueryType } from '../../../utils/textSearch';
import { DropDown } from './DropDown';
import _ from 'lodash';
import * as React from 'react';

export type ItemID = string | number;

export interface ShowItem {
  id: ItemID;
  label: string;
}

export interface AutocompleteProps<T> {
  onQueryChanged?: (query: QueryType) => void;
  onItemSelected?: (item: T) => void;
  minAcceptableLength?: number;
  debounceInterval?: number;
  placeholder?: string;
  maxOptionsToShow?: number;
  autoCalculate?: boolean;
  data: DataArrayType<T>;
  dataLabel?: (item: T) => string;
  dataId?: (item: T) => ItemID;
  caseSensitive?: boolean;
  showDataOnEmptyInput?: boolean;
  initialInput?: string;
  loading?: boolean;
  LoadingAnimation?: React.ComponentType<React.HTMLAttributes<HTMLElement>>;
  keepDropDownOnSelect?: boolean;
}
/*
 * This component renders input with autocompletion
 * It can be in two modes:
 * - autoCalculate ("static")
 *   it gets all the data once and then as user interacts, searches through this given data
 * - "dynamic"
 *   it assumes that all the calculations with data search by query happen outside of the component,
 *   and it just renders data
 */
export const Autocomplete = React.memo(
  <T,>({
    onQueryChanged,
    onItemSelected,
    minAcceptableLength = 3,
    debounceInterval = 600,
    placeholder = '',
    maxOptionsToShow = 5,
    autoCalculate,
    data = [],
    dataLabel = defaultLabelId,
    dataId = defaultLabelId,
    caseSensitive = false,
    showDataOnEmptyInput = true,
    initialInput,
    loading,
    LoadingAnimation,
    keepDropDownOnSelect,
  }: AutocompleteProps<T>) => {
    /**
     * If Autocomplete is used to display result of the search that were recalculated outside,
     * `autoCalculate` should be set to false. Then we just always need to convert each item to
     * pair `{id, label}`, and take first `maxOptionsToShow`.
     * So `showInit` should always be `true` when `autoCalculate` is false
     *
     * And only once user clicks on the input or focuses on it, this precalculated pairs should be
     * shown
     *
     * If Autocomplete is using static data and should filter it by itself as user types,
     * `showDataOnEmptyInput` lets us to choose, what to show when user didn't enter the input:
     * no options or just first `maxOptionsToShow` options
     **/

    // const [inputQueryState, setInput, setQuery] = useInputQuery(
    const [inputQueryState, setInput] = useInputQuery(debounceInterval, minAcceptableLength);
    const prevQuery = React.useRef<QueryType>(undefined);

    const [showDropDown, setShowDropDown] = React.useState(false);

    const itemsToShow = useDataShow<T>(
      data,
      dataLabel,
      dataId,
      maxOptionsToShow,
      autoCalculate,
      caseSensitive,
      inputQueryState.query
    );

    const inputRef = React.useRef<HTMLInputElement>(null);
    const dropDownRef = React.useRef<HTMLDivElement>(null);

    React.useEffect(() => {
      if (initialInput !== undefined) {
        setInput(initialInput || '');
      }
    }, [initialInput, setInput]);

    // listens to the query debounced, to make HTTP call for example
    // it should be called no matter if query changed due to user typing or user selected item,
    // we need to make sure external search result has all the correct data from server
    React.useEffect(() => {
      if (inputQueryState.query !== prevQuery.current) {
        prevQuery.current = inputQueryState.query;
        onQueryChanged?.(inputQueryState.query);
      }
    }, [inputQueryState.query, onQueryChanged]);

    // all the calculations of dataToShow should be made by this moment,
    // only show drop-down and select all tezt in input
    const onFocusIn = React.useCallback(() => {
      if (showDataOnEmptyInput || inputQueryState.input) {
        setShowDropDown(true);
        if (inputRef.current && inputQueryState.input) {
          inputRef.current.setSelectionRange(0, inputQueryState.input.length);
        }
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [dropDownRef?.current, inputRef?.current, inputQueryState.input, itemsToShow]);

    // listens to user typing
    const handleInputChange = React.useCallback(
      (e: React.ChangeEvent<HTMLInputElement>) => {
        const { value } = e.target;
        setInput(value);
      },
      [setInput]
    );

    // if user clicks outside of this component - close drop-down
    const onClickOutside = React.useCallback(() => {
      setShowDropDown(false);
    }, []);

    useClickOutside([dropDownRef, inputRef], onClickOutside);

    // listens on user selecting the option from drop-down
    const onSelectItem = React.useCallback(
      (item: DictionaryItem<T>) => {
        onItemSelected?.(item.itemData);
        keepDropDownOnSelect && setShowDropDown(false);
        setInput(item.itemToShow.label);
      },
      [onItemSelected, setInput, keepDropDownOnSelect]
    );

    return (
      <>
        <AutocompleteInput
          ref={inputRef}
          type="text"
          value={inputQueryState.input}
          onChange={handleInputChange}
          placeholder={placeholder}
          onFocus={onFocusIn}
        />

        {showDropDown && itemsToShow.length > 0 && (
          <DropDown
            ref={dropDownRef}
            onSelectItem={onSelectItem}
            items={itemsToShow}
            query={inputQueryState.query}
            caseSensitive={caseSensitive}
            loading={loading}
            LoadingAnimation={LoadingAnimation}
          />
        )}
      </>
    );
  },
  (prevProps: AutocompleteProps<any>, nextProps: AutocompleteProps<any>) => {
    const {
      dataLabel: prevDataLabel,
      dataId: prevDataId,
      onItemSelected: onPrevItemSelected,
      onQueryChanged: onPrevQueryChanged,
      ...prevRest
    } = prevProps;
    const {
      dataLabel: nextDataLabel,
      dataId: nextDataId,
      onItemSelected: onNextItemSelected,
      onQueryChanged: onNextQueryChanged,
      ...nextRest
    } = nextProps;
    return (
      _.isEqual(prevRest, nextRest) &&
      prevDataId === nextDataId &&
      prevDataLabel === nextDataLabel &&
      onPrevItemSelected === onNextItemSelected &&
      onPrevQueryChanged === onNextQueryChanged
    );
  }
);

Autocomplete.displayName = 'Autocomplete';
