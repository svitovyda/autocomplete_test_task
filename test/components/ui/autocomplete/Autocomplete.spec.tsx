import { Autocomplete } from '../../../../src/components/ui/autocomplete/Autocomplete';
import { MockListObjects, MockListStrings, toId, toLabel } from '../../../__mock__/mock';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import * as React from 'react';

describe('Autocomplete', () => {
  it('renders without error empty', () => {
    const component = render(<Autocomplete data={[]} />);

    expect(component).toMatchSnapshot();
  });

  it('renders without error strings', () => {
    const component = render(<Autocomplete data={MockListStrings} autoCalculate />);

    expect(component).toMatchSnapshot();
  });

  it('renders without error objects', () => {
    const component = render(
      <Autocomplete data={MockListObjects} autoCalculate dataId={toId} dataLabel={toLabel} />
    );

    expect(component).toMatchSnapshot();
  });
});
