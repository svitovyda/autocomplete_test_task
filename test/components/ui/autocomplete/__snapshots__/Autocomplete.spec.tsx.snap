// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete renders without error empty 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c0 {
  display: inline-block;
  padding: 4px 2px;
  cursor: text;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #aaa;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin: 0;
  padding: 8px;
  outline-width: 0;
  color: inherit;
}

.c0::-webkit-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::-moz-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0:-ms-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::placeholder {
  color: #aaa;
  font-size: 16px;
}

<body>
    <div>
      <input
        class="c0"
        placeholder=""
        type="text"
        value=""
      />
    </div>
  </body>,
  "container": .c0 {
  display: inline-block;
  padding: 4px 2px;
  cursor: text;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #aaa;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin: 0;
  padding: 8px;
  outline-width: 0;
  color: inherit;
}

.c0::-webkit-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::-moz-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0:-ms-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::placeholder {
  color: #aaa;
  font-size: 16px;
}

<div>
    <input
      class="c0"
      placeholder=""
      type="text"
      value=""
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Autocomplete renders without error objects 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c0 {
  display: inline-block;
  padding: 4px 2px;
  cursor: text;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #aaa;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin: 0;
  padding: 8px;
  outline-width: 0;
  color: inherit;
}

.c0::-webkit-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::-moz-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0:-ms-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::placeholder {
  color: #aaa;
  font-size: 16px;
}

<body>
    <div>
      <input
        class="c0"
        placeholder=""
        type="text"
        value=""
      />
    </div>
  </body>,
  "container": .c0 {
  display: inline-block;
  padding: 4px 2px;
  cursor: text;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #aaa;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin: 0;
  padding: 8px;
  outline-width: 0;
  color: inherit;
}

.c0::-webkit-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::-moz-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0:-ms-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::placeholder {
  color: #aaa;
  font-size: 16px;
}

<div>
    <input
      class="c0"
      placeholder=""
      type="text"
      value=""
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Autocomplete renders without error strings 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c0 {
  display: inline-block;
  padding: 4px 2px;
  cursor: text;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #aaa;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin: 0;
  padding: 8px;
  outline-width: 0;
  color: inherit;
}

.c0::-webkit-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::-moz-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0:-ms-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::placeholder {
  color: #aaa;
  font-size: 16px;
}

<body>
    <div>
      <input
        class="c0"
        placeholder=""
        type="text"
        value=""
      />
    </div>
  </body>,
  "container": .c0 {
  display: inline-block;
  padding: 4px 2px;
  cursor: text;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #aaa;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin: 0;
  padding: 8px;
  outline-width: 0;
  color: inherit;
}

.c0::-webkit-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::-moz-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0:-ms-input-placeholder {
  color: #aaa;
  font-size: 16px;
}

.c0::placeholder {
  color: #aaa;
  font-size: 16px;
}

<div>
    <input
      class="c0"
      placeholder=""
      type="text"
      value=""
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
